pkg/controller/podgc/gc_controller.go

```diff
@@ -124,6 +130,73 @@ func isPodTerminated(pod *v1.Pod) bool {
 	return false
 }
 
+// isPodTerminating returns true if the pod is terminating.
+func isPodTerminating(pod *v1.Pod) bool {
+	if pod.ObjectMeta.DeletionTimestamp != nil {
+		klog.Infof("garbage collecting pod %s that is terminating. Phase [%v]", pod.Name, pod.Status.Phase)
+		return true
+	}
+	return false
+}
+
+// podToleratesTaint returns true if the pod tolerates the provided taint.
+func podToleratesTaint(taint *v1.Taint, pod *v1.Pod) bool {
+	for _, toleration := range pod.Spec.Tolerations {
+		if toleration.ToleratesTaint(taint) {
+			return true
+		}
+	}
+	return false
+}
+
+func (gcc *PodGCController) gcTerminating(pods []*v1.Pod) {
+	klog.V(4).Info("GC'ing terminating pods that are on out of service nodes")
+	terminatingPods := []*v1.Pod{}
+	for _, pod := range pods {
+		if isPodTerminating(pod) {
+			node, err := gcc.nodeLister.Get(string(pod.Spec.NodeName))
+			if err != nil {
+				klog.Errorf("failed to get node:%s", err)
+				continue
+			}
+
+			// Add this pod to terminatingPods list only if the following conditions are met:
+			// 1. Node is not ready.
+			// 2. Node has `node.kubernetes.io/out-of-service` taint with `NoExecute` effect.
+			// 3. Pod does not tolerate the `node.kubernetes.io/out-of-service` taint with `NoExecute` effect.
+			taintToMatch := &v1.Taint{
+				Key:    v1.TaintNodeOutOfService,
+				Effect: v1.TaintEffectNoExecute,
+			}
+			if !nodeutil.IsNodeReady(node) && taints.TaintExists(node.Spec.Taints, taintToMatch) &&
+				!podToleratesTaint(taintToMatch, pod) {
+				terminatingPods = append(terminatingPods, pod)
+			}
+		}
+	}
+
+	deleteCount := len(terminatingPods)
+	if deleteCount == 0 {
+		return
+	}
+
+	klog.Infof("Garbage collecting %v pods that are terminating on node tainted with node.kubernetes.io/out-of-service", deleteCount)
+	// sort only when necessary
+	sort.Sort(byCreationTimestamp(terminatingPods))
+	var wait sync.WaitGroup
+	for i := 0; i < deleteCount; i++ {
+		wait.Add(1)
+		go func(namespace string, name string) {
+			defer wait.Done()
+			if err := gcc.deletePod(namespace, name); err != nil {
+				// ignore not founds
+				defer utilruntime.HandleError(err)
```

Comment:
Why are you deferring the error handling here? - It'll happen immediately regardless